tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/StreamingBlobStoreBackedReadHandleImpl.java

```diff
@@ -0,0 +1,283 @@
+        DataInputStream dataStream;
+
+        public GroupedReader(long ledgerId, long firstEntry, long lastEntry,
+                             StreamingOffloadIndexBlock index,
+                             BackedInputStream inputStream, DataInputStream dataStream) {
+            this.ledgerId = ledgerId;
+            this.firstEntry = firstEntry;
+            this.lastEntry = lastEntry;
+            this.index = index;
+            this.inputStream = inputStream;
+            this.dataStream = dataStream;
+        }
+    }
+
+    private StreamingBlobStoreBackedReadHandleImpl(long ledgerId, List<StreamingOffloadIndexBlock> indices,
+                                                   List<BackedInputStream> inputStreams,
+                                                   ExecutorService executor) {
+        this.ledgerId = ledgerId;
+        this.indices = indices;
+        this.inputStreams = inputStreams;
+        this.dataStreams = new LinkedList<>();
+        for (BackedInputStream inputStream : inputStreams) {
+            dataStreams.add(new DataInputStream(inputStream));
+        }
+        this.executor = executor;
+    }
+
+    @Override
+    public long getId() {
+        return ledgerId;
+    }
+
+    @Override
+    public LedgerMetadata getLedgerMetadata() {
+        return indices.get(0).getLedgerMetadata(ledgerId);
+    }
+
+    @Override
+    public CompletableFuture<Void> closeAsync() {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        executor.submit(() -> {
+            try {
+                for (StreamingOffloadIndexBlock indexBlock : indices) {
+                    indexBlock.close();
+                }
+                for (BackedInputStream inputStream : inputStreams) {
+                    inputStream.close();
+                }
+                promise.complete(null);
+            } catch (IOException t) {
+                promise.completeExceptionally(t);
+            }
+        });
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<LedgerEntries> readAsync(long firstEntry, long lastEntry) {
+        log.debug("Ledger {}: reading {} - {}", getId(), firstEntry, lastEntry);
+        CompletableFuture<LedgerEntries> promise = new CompletableFuture<>();
+        if (firstEntry > lastEntry
+                || firstEntry < 0
+                || lastEntry > getLastAddConfirmed()) {
+            promise.completeExceptionally(new BKException.BKIncorrectParameterException());
+            return promise;
+        }
+        executor.submit(() -> {
+            List<LedgerEntry> entries = new ArrayList<LedgerEntry>();
+            List<GroupedReader> groupedReaders = null;
+            try {
+                groupedReaders = getGroupedReader(firstEntry, lastEntry);
+            } catch (Exception e) {
+                promise.completeExceptionally(e);
+                return;
+            }
+
+            for (GroupedReader groupedReader : groupedReaders) {
```

Comment:
Please double confirm the `groupedReaders` are sorted by ledger Id, entry id, Otherwise, this will bread the entry read in order guarantee

Reply:
double confirm code added